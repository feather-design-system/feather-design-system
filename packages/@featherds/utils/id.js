const getSafeId = function (str) {
  str = str || "feather";
  const random = Math.floor(Math.random() * 1000000000);
  return [str.replace(/\s+/g, "-"), Date.now(), random].join("-");
};

export { getSafeId };
