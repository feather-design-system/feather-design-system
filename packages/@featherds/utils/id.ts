const getSafeId = function (str: string) {
  str = str || "feather";
  const random = Math.floor(Math.random() * 1000000000);
  return [str.replace(/\s+/g, "-"), Date.now(), random].join("-");
};

export { getSafeId };
