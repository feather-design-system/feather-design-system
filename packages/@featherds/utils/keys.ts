const KEYCODES = {
  PAGEUP: 33,
  PAGEDOWN: 34,
  UP: 38,
  DOWN: 40,
  LEFT: 37,
  RIGHT: 39,
  HOME: 36,
  END: 35,
  ENTER: 13,
  SPACE: 32,
  ESCAPE: 27,
  BACKSPACE: 8,
  DELETE: 46,
};

export { KEYCODES };
