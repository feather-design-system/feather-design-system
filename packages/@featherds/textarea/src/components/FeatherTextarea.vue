<template>
  <div v-bind="inherittedAttrs" class="feather-textarea-container">
    <InputWrapper
      :for="inputId"
      :raised="isRaised"
      :focused="focused"
      :show-clear="showClear"
      @wrapper-click="handleWrapperClick"
      @clear="handleClear"
      class="feather-input-content"
      :class="contentCls"
    >
      <textarea
        v-bind="attrs"
        class="feather-textarea"
        data-ref-id="feather-input"
        :class="{ error: error }"
        :maxlength="maxlength > 0 ? maxlength : false"
        ref="input"
      ></textarea>
    </InputWrapper>
    <InputSubText :id="descriptionId">
      <template v-slot:right>
        <div
          class="feather-input-count"
          v-if="maxlength"
          data-ref-id="feather-form-element-count"
        >
          {{ charCount }}
        </div>
      </template>
    </InputSubText>
  </div>
</template>
<script>
import { defineComponent } from "vue";
import { getSafeId } from "@featherds/utils/id";
import { useValidation } from "@featherds/input/src/components/useValidation";
import { computed, toRef } from "vue";

import {
  InputWrapper,
  InputWrapperMixin,
  InputSubText,
  InputSubTextMixin,
  InputInheritAttrsMixin,
} from "@featherds/input-helper";
export default defineComponent({
  model: {
    prop: "modelValue",
    event: "update:modelValue",
  },
  emits: ["update:modelValue"],
  mixins: [InputWrapperMixin, InputSubTextMixin, InputInheritAttrsMixin],
  props: {
    modelValue: {
      type: String,
    },
    maxlength: {
      type: Number,
      required: false,
      default: 0,
    },
    auto: {
      type: Boolean,
      default: false,
    },
    schema: {
      type: Object,
      required: false,
    },
  },
  data() {
    return {
      focused: false,
      internalValue: undefined,
      initialHeight: 0,
    };
  },
  computed: {
    descriptionId() {
      return getSafeId("feather-textarea-description");
    },
    showClear() {
      return !!(this.internalValue && this.internalValue.length > 0);
    },
    contentCls() {
      const cls = [];

      if (this.error) {
        cls.push("error");
      }

      if (this.disabled) {
        cls.push("disabled");
      }
      if (this.focused) {
        cls.push("focused");
      }
      return cls;
    },
    isRaised() {
      //is a string with length 0 return false
      if (this.internalValue || this.focused) {
        return true;
      }
      return false;
    },

    attrs() {
      const attrs = { ...this.$attrs };
      delete attrs.placeholder;
      //use aria-invalid if passed in (some validation libraries will specify this)
      //otherwise base it off this.error
      if (!attrs["aria-invalid"]) {
        attrs["aria-invalid"] = !!this.error;
      }
      return {
        ...attrs,
        ...this.listeners,
        class: "",
        id: this.inputId,
        name: this.inputId,
        disabled: this.disabled,
        "aria-disabled": this.disabled,
        "aria-describedby": (attrs["aria-describedby"] || "")
          .split(" ")
          .concat([this.descriptionId])
          .filter((x) => x.length > 0)
          .join(" "),
        value: this.internalValue,
      };
    },
    listeners() {
      return {
        onFocus: (e) => {
          this.handleFocus();
          if (this.$attrs.onFocus) {
            this.$attrs.onFocus(e);
          }
        },
        onBlur: (e) => {
          this.handleBlur();
          if (this.$attrs.onBlur) {
            this.$attrs.onBlur(e);
          }
        },
        onInput: (e) => {
          this.adjustTextArea();
          this.handleInput(e);
        },
      };
    },
    charCount() {
      return `${(this.internalValue && this.internalValue.length) || "0"} / ${
        this.maxlength
      }`;
    },
  },
  watch: {
    modelValue: {
      immediate: true,
      handler(v) {
        this.adjustTextArea();
        this.internalValue = v;
      },
    },
    internalValue: {
      immediate: true,
      handler(v) {
        this.$emit("update:modelValue", v);
      },
    },
  },
  setup(props) {
    const incomingId = toRef(props, "id");
    const inputId = computed(() => {
      if (incomingId.value) {
        return incomingId.value;
      }
      return getSafeId("feather-textarea-label");
    });

    const { validate } = useValidation(
      inputId,
      toRef(props, "modelValue"),
      props.label,
      props.schema
    );

    return { inputId, incomingId, validate };
  },
  methods: {
    handleClear() {
      this.internalValue = "";
      this.focus();
    },
    handleWrapperClick() {
      this.$refs.input.focus();
    },
    handleFocus() {
      this.focused = true;
    },
    handleBlur() {
      this.validate();
      this.focused = false;
    },
    handleInput(e) {
      this.internalValue = e.target.value;
      this.$emit("update:modelValue", this.internalValue);
    },
    focus() {
      this.$nextTick(() => {
        this.$refs.input.focus();
      });
    },
    adjustTextArea() {
      if (!this.auto) {
        return;
      }
      //auto adjust height
      const tx = this.$refs.input;
      //if this is called before mount
      if (!tx) {
        return;
      }
      tx.style.height = "12px";
      tx.style.flexBasis = "40px";
      tx.style.whiteSpace = "nowrap";

      this.$nextTick(() => {
        const txWidth = tx.getBoundingClientRect().width;
        if (
          tx.scrollWidth <= tx.clientWidth &&
          txWidth < tx.parentElement.getBoundingClientRect().width
        ) {
          tx.style.whiteSpace = "nowrap";
        } else {
          tx.style.whiteSpace = "normal";
          tx.style.flexBasis = "100%";
        }
        this.$nextTick(() => {
          tx.style.height = `${
            tx.scrollHeight < this.initialHeight
              ? this.initialHeight
              : tx.scrollHeight
          }px`;
        });
      });
    },
  },
  mounted() {
    const tx = this.$refs.input;
    this.initialHeight = tx.getBoundingClientRect().height;
  },
  components: {
    InputSubText,
    InputWrapper,
  },
});
</script>

<style lang="scss" scoped>
@import "@featherds/styles/themes/variables";
@import "@featherds/styles/mixins/typography";
@import "@featherds/input-helper/scss/spacing";
.feather-textarea-container {
  @include input-spacing;
}
.feather-textarea {
  border: none;
  margin: 7px 0;
  display: inline-block;
  vertical-align: middle;
  resize: none;
  background-color: transparent;
  flex: 1 1 40px;
  width: 0;
  @include body-small();
  line-height: 1.5em; //IE has an issue with the mixin lineheight when typing
  color: var($primary-text-on-surface);
  caret-color: var($primary);
  overflow: auto;
  &::-ms-clear {
    display: none;
  }
  &:focus {
    outline: 0;
  }
  &.error {
    caret-color: var($error);
  }
  &[disabled] {
    color: var($disabled-text-on-surface);
  }
}

.feather-input-count {
  padding-left: 16px;
  align-self: flex-start;
  margin-left: auto;
  @include caption();
  color: var($secondary-text-on-surface);
}

.feather-textarea-container {
  :deep(.prefix),
  :deep(.post) {
    align-self: flex-start;
    margin-top: 10px;
  }
}
</style>
