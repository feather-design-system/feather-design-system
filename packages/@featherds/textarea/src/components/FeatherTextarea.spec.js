import { nextTick } from "vue";
import axe from "@featherds/utils/test/axe";
import FeatherTextarea from "./FeatherTextarea.vue";
import { mount, config } from "@vue/test-utils";
import * as id from "@featherds/utils/id";
import "@featherds/input-helper/test/MutationObserver";

config.renderStubDefaultSlot = true;

jest.spyOn(id, "getSafeId").mockImplementation((x) => x);

const getWrapper = function (options) {
  return mount(FeatherTextarea, options);
};

describe("FeatherTextarea.vue", () => {
  it("should render an textarea with a label", () => {
    const wrapper = getWrapper({
      props: {
        label: "label",
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should render an textarea with a label and hint text", () => {
    const wrapper = getWrapper({
      props: {
        label: "label",
        hint: "hint text",
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should render an textarea with a preset value", () => {
    const wrapper = getWrapper({
      props: {
        label: "label",
        hint: "hint text",
        modelValue: "YEET",
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should render an textarea with raised label when focused", async () => {
    const wrapper = getWrapper({
      props: {
        label: "label",
      },
    });
    await wrapper.find("textarea").trigger("focus");
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should render an textarea with raised label when focused and has a value", async () => {
    const wrapper = getWrapper({
      props: {
        label: "label",
        modelValue: "YEET",
      },
    });
    await wrapper.find("textarea").trigger("focus");
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should trigger textarea event when something is typed", () => {
    const wrapper = getWrapper({
      props: {
        label: "label",
      },
    });
    const textarea = wrapper.find("textarea");
    const expected = "TEEST";
    textarea.element.value = expected;
    textarea.trigger("input");
    expect(wrapper.emitted("update:modelValue")[1][0]).toBe(expected);
  });

  it("should render an textarea error", () => {
    const wrapper = getWrapper({
      props: {
        label: "label",
        error: "Error text",
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should render an textarea error and has focus", async () => {
    const wrapper = getWrapper({
      props: {
        label: "label",
        error: "Error text",
      },
    });
    await wrapper.find("textarea").trigger("focus");
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should render an textarea error and has a value", () => {
    const wrapper = getWrapper({
      props: {
        label: "label",
        error: "Error text",
        modelValue: "Test",
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should render a clear icon when clear has a value and textarea also has a value", () => {
    const wrapper = getWrapper({
      props: {
        label: "label",
        clear: "clear",
        modelValue: "Test",
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should not render a clear icon when the textarea doesn't have a value", () => {
    const wrapper = getWrapper({
      props: {
        label: "label",
        clear: "clear",
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should render a maxlength attribute and character count when maxlength specified", () => {
    const wrapper = getWrapper({
      props: {
        label: "label",
        modelValue: "test",
        maxlength: 5,
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should clear textarea when clear-icon is clicked", async () => {
    const wrapper = getWrapper({
      props: {
        label: "label",
        modelValue: "YEET",
        clear: "clear",
      },
    });
    wrapper.vm.handleClear();
    await nextTick();
    expect(wrapper.element).toMatchSnapshot();
    expect(wrapper.emitted("update:modelValue")[1][0]).toBe("");
  });

  it("should allow custom focus events", async () => {
    const fn = jest.fn();
    const wrapper = getWrapper({
      props: {
        label: "label",
        modelValue: "YEET",
      },
      attrs: {
        onFocus: fn,
      },
    });

    await wrapper.find("textarea").trigger("focus");
    expect(fn).toHaveBeenCalled();
    expect(wrapper.element).toMatchSnapshot();
  });

  fit("should allow custom blur events", () => {
    const fn = jest.fn();
    const wrapper = getWrapper({
      props: {
        label: "label",
        modelValue: "YEET",
      },
      attrs: {
        onBlur: fn,
      },
    });

    wrapper.find("textarea").trigger("blur");

    expect(fn).toHaveBeenCalled();
    expect(wrapper.element).toMatchSnapshot();
  });

  describe("a11y", () => {
    const accessibilityTest = async (options) => {
      const wrapper = mount(FeatherTextarea, options);
      expect(await axe(wrapper.element)).toHaveNoViolations();
    };
    it("should be accessible", async () => {
      await accessibilityTest({
        props: {
          label: "label",
          modelValue: "YEET",
        },
      });
    });
    it("should be accessible with hint", async () => {
      await accessibilityTest({
        props: {
          label: "label",
          modelValue: "YEET",
          hint: "HINTTT",
        },
      });
    });
    it("should be accessible with error", async () => {
      await accessibilityTest({
        props: {
          label: "label",
          modelValue: "YEET",
          error: "Error",
        },
      });
    });
  });
});
