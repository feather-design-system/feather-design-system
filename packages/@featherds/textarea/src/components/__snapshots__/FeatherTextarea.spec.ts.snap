// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeatherTextarea.vue should allow custom blur events 1`] = `
<div
  class="feather-textarea-container"
>
  <div
    class="feather-input-wrapper-container raised feather-textarea-content"
  >
    <div
      class="feather-input-border"
    >
      <div
        class="pre-border"
      />
      <div
        class="label-border"
        style="width: 0px;"
      >
        <label
          class="feather-input-label"
          data-ref-id="feather-form-element-label"
          for="feather-textarea-label"
        >
          label
        </label>
      </div>
      <div
        class="post-border"
      />
    </div>
    <div
      class="feather-input-wrapper"
    >
      <div
        class="prefix"
      >
        
        
      </div>
      
      <textarea
        aria-describedby="feather-textarea-description"
        aria-disabled="false"
        aria-invalid="false"
        class="feather-textarea"
        data-ref-id="feather-textarea-input"
        id="feather-textarea-label"
        name="feather-textarea-label"
      />
      
      <div
        class="post"
      >
        <!--v-if-->
        <!--v-if-->
        
        
      </div>
    </div>
  </div>
  <div
    class="feather-input-sub-text"
    id="feather-textarea-description"
  >
    <div
      class="feather-input-spacer"
    >
       
    </div>
    <!--v-if-->
    <!--v-if-->
    
    
  </div>
</div>
`;

exports[`FeatherTextarea.vue should allow custom focus events 1`] = `
<div
  class="feather-textarea-container"
>
  <div
    class="feather-input-wrapper-container raised focused feather-textarea-content focused"
  >
    <div
      class="feather-input-border"
    >
      <div
        class="pre-border"
      />
      <div
        class="label-border"
        style="width: 0px;"
      >
        <label
          class="feather-input-label"
          data-ref-id="feather-form-element-label"
          for="feather-textarea-label"
        >
          label
        </label>
      </div>
      <div
        class="post-border"
      />
    </div>
    <div
      class="feather-input-wrapper"
    >
      <div
        class="prefix"
      >
        
        
      </div>
      
      <textarea
        aria-describedby="feather-textarea-description"
        aria-disabled="false"
        aria-invalid="false"
        class="feather-textarea"
        data-ref-id="feather-textarea-input"
        id="feather-textarea-label"
        name="feather-textarea-label"
      />
      
      <div
        class="post"
      >
        <!--v-if-->
        <!--v-if-->
        
        
      </div>
    </div>
  </div>
  <div
    class="feather-input-sub-text"
    id="feather-textarea-description"
  >
    <div
      class="feather-input-spacer"
    >
       
    </div>
    <!--v-if-->
    <!--v-if-->
    
    
  </div>
</div>
`;

exports[`FeatherTextarea.vue should clear textarea when clear-icon is clicked 1`] = `
<div
  class="feather-textarea-container"
>
  <div
    class="feather-input-wrapper-container feather-textarea-content"
  >
    <div
      class="feather-input-border"
    >
      <div
        class="pre-border"
      />
      <div
        class="label-border"
        style="width: 0px;"
      >
        <label
          class="feather-input-label"
          data-ref-id="feather-form-element-label"
          for="feather-textarea-label"
        >
          label
        </label>
      </div>
      <div
        class="post-border"
      />
    </div>
    <div
      class="feather-input-wrapper"
    >
      <div
        class="prefix"
      >
        
        
      </div>
      
      <textarea
        aria-describedby="feather-textarea-description"
        aria-disabled="false"
        aria-invalid="false"
        class="feather-textarea"
        data-ref-id="feather-textarea-input"
        id="feather-textarea-label"
        name="feather-textarea-label"
      />
      
      <div
        class="post"
      >
        <!--v-if-->
        <!--v-if-->
        
        
      </div>
    </div>
  </div>
  <div
    class="feather-input-sub-text"
    id="feather-textarea-description"
  >
    <div
      class="feather-input-spacer"
    >
       
    </div>
    <!--v-if-->
    <!--v-if-->
    
    
  </div>
</div>
`;

exports[`FeatherTextarea.vue should not render a clear icon when the textarea doesn't have a value 1`] = `
<div
  class="feather-textarea-container"
>
  <div
    class="feather-input-wrapper-container feather-textarea-content"
  >
    <div
      class="feather-input-border"
    >
      <div
        class="pre-border"
      />
      <div
        class="label-border"
        style="width: 0px;"
      >
        <label
          class="feather-input-label"
          data-ref-id="feather-form-element-label"
          for="feather-textarea-label"
        >
          label
        </label>
      </div>
      <div
        class="post-border"
      />
    </div>
    <div
      class="feather-input-wrapper"
    >
      <div
        class="prefix"
      >
        
        
      </div>
      
      <textarea
        aria-describedby="feather-textarea-description"
        aria-disabled="false"
        aria-invalid="false"
        class="feather-textarea"
        data-ref-id="feather-textarea-input"
        id="feather-textarea-label"
        name="feather-textarea-label"
      />
      
      <div
        class="post"
      >
        <!--v-if-->
        <!--v-if-->
        
        
      </div>
    </div>
  </div>
  <div
    class="feather-input-sub-text"
    id="feather-textarea-description"
  >
    <div
      class="feather-input-spacer"
    >
       
    </div>
    <!--v-if-->
    <!--v-if-->
    
    
  </div>
</div>
`;

exports[`FeatherTextarea.vue should render a clear icon when clear has a value and textarea also has a value 1`] = `
<div
  class="feather-textarea-container"
>
  <div
    class="feather-input-wrapper-container raised feather-textarea-content"
  >
    <div
      class="feather-input-border"
    >
      <div
        class="pre-border"
      />
      <div
        class="label-border"
        style="width: 0px;"
      >
        <label
          class="feather-input-label"
          data-ref-id="feather-form-element-label"
          for="feather-textarea-label"
        >
          label
        </label>
      </div>
      <div
        class="post-border"
      />
    </div>
    <div
      class="feather-input-wrapper"
    >
      <div
        class="prefix"
      >
        
        
      </div>
      
      <textarea
        aria-describedby="feather-textarea-description"
        aria-disabled="false"
        aria-invalid="false"
        class="feather-textarea"
        data-ref-id="feather-textarea-input"
        id="feather-textarea-label"
        name="feather-textarea-label"
      />
      
      <div
        class="post"
      >
        <a
          class="action-icon hide-when-disabled"
          data-ref-id="feather-form-element-clear"
          href="#"
          title="clear"
        >
          <svg
            aria-hidden="true"
            class="feather-icon"
            focusable="false"
            role="img"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M19,17.55,13.43,12,19,6.4A1,1,0,1,0,17.58,5L12,10.58,6.42,5A1,1,0,0,0,5,6.42L10.6,12,5,17.6A1,1,0,0,0,5,19a1,1,0,0,0,.71.29,1,1,0,0,0,.71-.3L12,13.41,17.6,19a1,1,0,0,0,.71.29A1,1,0,0,0,19,19,1,1,0,0,0,19,17.55Z"
            />
          </svg>
        </a>
        <!--v-if-->
        
        
      </div>
    </div>
  </div>
  <div
    class="feather-input-sub-text"
    id="feather-textarea-description"
  >
    <div
      class="feather-input-spacer"
    >
       
    </div>
    <!--v-if-->
    <!--v-if-->
    
    
  </div>
</div>
`;

exports[`FeatherTextarea.vue should render a maxlength attribute and character count when maxlength specified 1`] = `
<div
  class="feather-textarea-container"
>
  <div
    class="feather-input-wrapper-container raised feather-textarea-content"
  >
    <div
      class="feather-input-border"
    >
      <div
        class="pre-border"
      />
      <div
        class="label-border"
        style="width: 0px;"
      >
        <label
          class="feather-input-label"
          data-ref-id="feather-form-element-label"
          for="feather-textarea-label"
        >
          label
        </label>
      </div>
      <div
        class="post-border"
      />
    </div>
    <div
      class="feather-input-wrapper"
    >
      <div
        class="prefix"
      >
        
        
      </div>
      
      <textarea
        aria-describedby="feather-textarea-description"
        aria-disabled="false"
        aria-invalid="false"
        class="feather-textarea"
        data-ref-id="feather-textarea-input"
        id="feather-textarea-label"
        maxlength="5"
        name="feather-textarea-label"
      />
      
      <div
        class="post"
      >
        <!--v-if-->
        <!--v-if-->
        
        
      </div>
    </div>
  </div>
  <div
    class="feather-input-sub-text"
    id="feather-textarea-description"
  >
    <div
      class="feather-input-spacer"
    >
       
    </div>
    <!--v-if-->
    <!--v-if-->
    
    <div
      class="feather-textarea-count"
      data-ref-id="feather-form-element-count"
    >
      4 / 5
    </div>
    
  </div>
</div>
`;

exports[`FeatherTextarea.vue should render an textarea error 1`] = `
<div
  class="feather-textarea-container"
>
  <div
    class="feather-input-wrapper-container error feather-textarea-content error"
  >
    <div
      class="feather-input-border"
    >
      <div
        class="pre-border"
      />
      <div
        class="label-border"
        style="width: 0px;"
      >
        <label
          class="feather-input-label"
          data-ref-id="feather-form-element-label"
          for="feather-textarea-label"
        >
          label
        </label>
      </div>
      <div
        class="post-border"
      />
    </div>
    <div
      class="feather-input-wrapper"
    >
      <div
        class="prefix"
      >
        
        
      </div>
      
      <textarea
        aria-describedby="feather-textarea-description"
        aria-disabled="false"
        aria-invalid="true"
        class="feather-textarea error"
        data-ref-id="feather-textarea-input"
        id="feather-textarea-label"
        name="feather-textarea-label"
      />
      
      <div
        class="post"
      >
        <!--v-if-->
        <svg
          aria-hidden="true"
          class="error-icon hide-when-disabled feather-icon error-icon hide-when-disabled"
          data-ref-id="error-icon"
          focusable="false"
          role="img"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12,2A10,10,0,1,0,22,12,10,10,0,0,0,12,2Zm0,18a8,8,0,1,1,8-8A8,8,0,0,1,12,20Z"
          />
          <rect
            height="7"
            rx="1"
            width="2"
            x="11"
            y="7"
          />
          <rect
            height="2"
            rx="0.65"
            width="2"
            x="11"
            y="15"
          />
        </svg>
        
        
      </div>
    </div>
  </div>
  <div
    class="feather-input-sub-text"
    id="feather-textarea-description"
  >
    <!--v-if-->
    <!--v-if-->
    <div
      aria-live="assertive"
      class="feather-input-error"
      data-ref-id="feather-form-element-error"
    >
      Error text
    </div>
    
    
  </div>
</div>
`;

exports[`FeatherTextarea.vue should render an textarea error and has a value 1`] = `
<div
  class="feather-textarea-container"
>
  <div
    class="feather-input-wrapper-container raised error feather-textarea-content error"
  >
    <div
      class="feather-input-border"
    >
      <div
        class="pre-border"
      />
      <div
        class="label-border"
        style="width: 0px;"
      >
        <label
          class="feather-input-label"
          data-ref-id="feather-form-element-label"
          for="feather-textarea-label"
        >
          label
        </label>
      </div>
      <div
        class="post-border"
      />
    </div>
    <div
      class="feather-input-wrapper"
    >
      <div
        class="prefix"
      >
        
        
      </div>
      
      <textarea
        aria-describedby="feather-textarea-description"
        aria-disabled="false"
        aria-invalid="true"
        class="feather-textarea error"
        data-ref-id="feather-textarea-input"
        id="feather-textarea-label"
        name="feather-textarea-label"
      />
      
      <div
        class="post"
      >
        <!--v-if-->
        <svg
          aria-hidden="true"
          class="error-icon hide-when-disabled feather-icon error-icon hide-when-disabled"
          data-ref-id="error-icon"
          focusable="false"
          role="img"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12,2A10,10,0,1,0,22,12,10,10,0,0,0,12,2Zm0,18a8,8,0,1,1,8-8A8,8,0,0,1,12,20Z"
          />
          <rect
            height="7"
            rx="1"
            width="2"
            x="11"
            y="7"
          />
          <rect
            height="2"
            rx="0.65"
            width="2"
            x="11"
            y="15"
          />
        </svg>
        
        
      </div>
    </div>
  </div>
  <div
    class="feather-input-sub-text"
    id="feather-textarea-description"
  >
    <!--v-if-->
    <!--v-if-->
    <div
      aria-live="assertive"
      class="feather-input-error"
      data-ref-id="feather-form-element-error"
    >
      Error text
    </div>
    
    
  </div>
</div>
`;

exports[`FeatherTextarea.vue should render an textarea error and has focus 1`] = `
<div
  class="feather-textarea-container"
>
  <div
    class="feather-input-wrapper-container raised focused error feather-textarea-content error focused"
  >
    <div
      class="feather-input-border"
    >
      <div
        class="pre-border"
      />
      <div
        class="label-border"
        style="width: 0px;"
      >
        <label
          class="feather-input-label"
          data-ref-id="feather-form-element-label"
          for="feather-textarea-label"
        >
          label
        </label>
      </div>
      <div
        class="post-border"
      />
    </div>
    <div
      class="feather-input-wrapper"
    >
      <div
        class="prefix"
      >
        
        
      </div>
      
      <textarea
        aria-describedby="feather-textarea-description"
        aria-disabled="false"
        aria-invalid="true"
        class="feather-textarea error"
        data-ref-id="feather-textarea-input"
        id="feather-textarea-label"
        name="feather-textarea-label"
      />
      
      <div
        class="post"
      >
        <!--v-if-->
        <svg
          aria-hidden="true"
          class="error-icon hide-when-disabled feather-icon error-icon hide-when-disabled"
          data-ref-id="error-icon"
          focusable="false"
          role="img"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12,2A10,10,0,1,0,22,12,10,10,0,0,0,12,2Zm0,18a8,8,0,1,1,8-8A8,8,0,0,1,12,20Z"
          />
          <rect
            height="7"
            rx="1"
            width="2"
            x="11"
            y="7"
          />
          <rect
            height="2"
            rx="0.65"
            width="2"
            x="11"
            y="15"
          />
        </svg>
        
        
      </div>
    </div>
  </div>
  <div
    class="feather-input-sub-text"
    id="feather-textarea-description"
  >
    <!--v-if-->
    <!--v-if-->
    <div
      aria-live="assertive"
      class="feather-input-error"
      data-ref-id="feather-form-element-error"
    >
      Error text
    </div>
    
    
  </div>
</div>
`;

exports[`FeatherTextarea.vue should render an textarea with a label 1`] = `
<div
  class="feather-textarea-container"
>
  <div
    class="feather-input-wrapper-container feather-textarea-content"
  >
    <div
      class="feather-input-border"
    >
      <div
        class="pre-border"
      />
      <div
        class="label-border"
        style="width: 0px;"
      >
        <label
          class="feather-input-label"
          data-ref-id="feather-form-element-label"
          for="feather-textarea-label"
        >
          label
        </label>
      </div>
      <div
        class="post-border"
      />
    </div>
    <div
      class="feather-input-wrapper"
    >
      <div
        class="prefix"
      >
        
        
      </div>
      
      <textarea
        aria-describedby="feather-textarea-description"
        aria-disabled="false"
        aria-invalid="false"
        class="feather-textarea"
        data-ref-id="feather-textarea-input"
        id="feather-textarea-label"
        name="feather-textarea-label"
      />
      
      <div
        class="post"
      >
        <!--v-if-->
        <!--v-if-->
        
        
      </div>
    </div>
  </div>
  <div
    class="feather-input-sub-text"
    id="feather-textarea-description"
  >
    <div
      class="feather-input-spacer"
    >
       
    </div>
    <!--v-if-->
    <!--v-if-->
    
    
  </div>
</div>
`;

exports[`FeatherTextarea.vue should render an textarea with a label and hint text 1`] = `
<div
  class="feather-textarea-container"
>
  <div
    class="feather-input-wrapper-container feather-textarea-content"
  >
    <div
      class="feather-input-border"
    >
      <div
        class="pre-border"
      />
      <div
        class="label-border"
        style="width: 0px;"
      >
        <label
          class="feather-input-label"
          data-ref-id="feather-form-element-label"
          for="feather-textarea-label"
        >
          label
        </label>
      </div>
      <div
        class="post-border"
      />
    </div>
    <div
      class="feather-input-wrapper"
    >
      <div
        class="prefix"
      >
        
        
      </div>
      
      <textarea
        aria-describedby="feather-textarea-description"
        aria-disabled="false"
        aria-invalid="false"
        class="feather-textarea"
        data-ref-id="feather-textarea-input"
        id="feather-textarea-label"
        name="feather-textarea-label"
      />
      
      <div
        class="post"
      >
        <!--v-if-->
        <!--v-if-->
        
        
      </div>
    </div>
  </div>
  <div
    class="feather-input-sub-text"
    id="feather-textarea-description"
  >
    <!--v-if-->
    <div
      class="feather-input-hint"
      data-ref-id="feather-form-element-hint"
    >
      hint text
    </div>
    <!--v-if-->
    
    
  </div>
</div>
`;

exports[`FeatherTextarea.vue should render an textarea with a preset value 1`] = `
<div
  class="feather-textarea-container"
>
  <div
    class="feather-input-wrapper-container raised feather-textarea-content"
  >
    <div
      class="feather-input-border"
    >
      <div
        class="pre-border"
      />
      <div
        class="label-border"
        style="width: 0px;"
      >
        <label
          class="feather-input-label"
          data-ref-id="feather-form-element-label"
          for="feather-textarea-label"
        >
          label
        </label>
      </div>
      <div
        class="post-border"
      />
    </div>
    <div
      class="feather-input-wrapper"
    >
      <div
        class="prefix"
      >
        
        
      </div>
      
      <textarea
        aria-describedby="feather-textarea-description"
        aria-disabled="false"
        aria-invalid="false"
        class="feather-textarea"
        data-ref-id="feather-textarea-input"
        id="feather-textarea-label"
        name="feather-textarea-label"
      />
      
      <div
        class="post"
      >
        <!--v-if-->
        <!--v-if-->
        
        
      </div>
    </div>
  </div>
  <div
    class="feather-input-sub-text"
    id="feather-textarea-description"
  >
    <!--v-if-->
    <div
      class="feather-input-hint"
      data-ref-id="feather-form-element-hint"
    >
      hint text
    </div>
    <!--v-if-->
    
    
  </div>
</div>
`;

exports[`FeatherTextarea.vue should render an textarea with raised label when focused 1`] = `
<div
  class="feather-textarea-container"
>
  <div
    class="feather-input-wrapper-container raised focused feather-textarea-content focused"
  >
    <div
      class="feather-input-border"
    >
      <div
        class="pre-border"
      />
      <div
        class="label-border"
        style="width: 0px;"
      >
        <label
          class="feather-input-label"
          data-ref-id="feather-form-element-label"
          for="feather-textarea-label"
        >
          label
        </label>
      </div>
      <div
        class="post-border"
      />
    </div>
    <div
      class="feather-input-wrapper"
    >
      <div
        class="prefix"
      >
        
        
      </div>
      
      <textarea
        aria-describedby="feather-textarea-description"
        aria-disabled="false"
        aria-invalid="false"
        class="feather-textarea"
        data-ref-id="feather-textarea-input"
        id="feather-textarea-label"
        name="feather-textarea-label"
      />
      
      <div
        class="post"
      >
        <!--v-if-->
        <!--v-if-->
        
        
      </div>
    </div>
  </div>
  <div
    class="feather-input-sub-text"
    id="feather-textarea-description"
  >
    <div
      class="feather-input-spacer"
    >
       
    </div>
    <!--v-if-->
    <!--v-if-->
    
    
  </div>
</div>
`;

exports[`FeatherTextarea.vue should render an textarea with raised label when focused and has a value 1`] = `
<div
  class="feather-textarea-container"
>
  <div
    class="feather-input-wrapper-container raised focused feather-textarea-content focused"
  >
    <div
      class="feather-input-border"
    >
      <div
        class="pre-border"
      />
      <div
        class="label-border"
        style="width: 0px;"
      >
        <label
          class="feather-input-label"
          data-ref-id="feather-form-element-label"
          for="feather-textarea-label"
        >
          label
        </label>
      </div>
      <div
        class="post-border"
      />
    </div>
    <div
      class="feather-input-wrapper"
    >
      <div
        class="prefix"
      >
        
        
      </div>
      
      <textarea
        aria-describedby="feather-textarea-description"
        aria-disabled="false"
        aria-invalid="false"
        class="feather-textarea"
        data-ref-id="feather-textarea-input"
        id="feather-textarea-label"
        name="feather-textarea-label"
      />
      
      <div
        class="post"
      >
        <!--v-if-->
        <!--v-if-->
        
        
      </div>
    </div>
  </div>
  <div
    class="feather-input-sub-text"
    id="feather-textarea-description"
  >
    <div
      class="feather-input-spacer"
    >
       
    </div>
    <!--v-if-->
    <!--v-if-->
    
    
  </div>
</div>
`;
